[JSC] Make a bit defensive on jettisoning

Constellation opened the pull request:
#### fc1a7277fce0dbb99daa15afae137a5188c106bb
<pre>
[JSC] Make a bit defensive on jettisoning
<a href="https://bugs.webkit.org/show_bug.cgi?id=274672">https://bugs.webkit.org/show_bug.cgi?id=274672</a>
<a href="https://rdar.apple.com/126591959">rdar://126591959</a>

Reviewed by Yijia Huang.

Speculatively making a bit defensive against jettisoning.

1. Ensure jettisoning does not add dead CodeBlocks.
2. Loop finalizeCodeBlockEdge until we ensure that edge is cleared or edge is alive.
3. Clear CallSlot for dead cells in PolymorphicCallStubRoutine (not necessary, but just for defensive change).

* Source/JavaScriptCore/jit/PolymorphicCallStubRoutine.cpp:
(JSC::PolymorphicCallStubRoutine::visitWeakImpl):
* Source/JavaScriptCore/runtime/ScriptExecutable.cpp:
(JSC::ScriptExecutable::installCode):
* Source/JavaScriptCore/runtime/ScriptExecutableInlines.h:
(JSC::ScriptExecutable::finalizeCodeBlockEdge):

Canonical link: <a href="https://commits.webkit.org/279292@main">https://commits.webkit.org/279292@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/90f01da8565d5a7d8a9169d83e07cb858e46ff97

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53025 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32362 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5512 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56304 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3748 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39205 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3474 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/2419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55123 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/30118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45772 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/24135 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3092 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1907 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/46381 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/48991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3249 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57899 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/52538 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28166 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3212 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29386 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45989 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/49709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/30305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/64843 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7795 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29140 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12306 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/runtime/ScriptExecutableInlines.h:
```diff
@@ -32,17 +32,23 @@ namespace JSC {
             codeBlock->jettison(Profiler::JettisonDueToOldAge);
-        if (codeBlock == codeBlockEdge.get())
+
+        if (codeBlock == codeBlockEdge.get()) {
```
Not really understand the logic here. So, after 
```
if (vm.heap.isMarked(codeBlock))
    return;
```
we know that the `codeBlock` is not alive. Then, we do jettison. After that we check whether the `codeBlockEdge` is already cleared or not. If not then clear it. So, why we don't do this?
```
if (codeBlockEdge.get()) { ... }
```

Constellation replied:
it is possible that jettisoning and related code will reinstall different code

hyjorc1 replied:
Maybe this is ensuring that the current `codeBlockEdge` is still valid and it's holding the current jettisoned `codeBlock`. Then, clear it.

And the `for` loop seems telling us that if at this point the `codeBlockEdge` may hold a nullptr then we still can start over to check the nullability and return.

hyjorc1 replied:
> it is possible that jettisoning and related code will reinstall different code

I see. In that case, just start over again and check whether the reinstalled code is alive or jettisoned. Make senses.

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Make a bit defensive on jettisoning
https://bugs.webkit.org/show_bug.cgi?id=274672
rdar://126591959

Reviewed by Yijia Huang.

Speculatively making a bit defensive against jettisoning.

1. Ensure jettisoning does not add dead CodeBlocks.
2. Loop finalizeCodeBlockEdge until we ensure that edge is cleared or edge is alive.
3. Clear CallSlot for dead cells in PolymorphicCallStubRoutine (not necessary, but just for defensive change).

* Source/JavaScriptCore/jit/PolymorphicCallStubRoutine.cpp:
(JSC::PolymorphicCallStubRoutine::visitWeakImpl):
* Source/JavaScriptCore/runtime/ScriptExecutable.cpp:
(JSC::ScriptExecutable::installCode):
* Source/JavaScriptCore/runtime/ScriptExecutableInlines.h:
(JSC::ScriptExecutable::finalizeCodeBlockEdge):

Canonical link: https://commits.webkit.org/279292@main
